compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -3689,12 +3689,10 @@ private boolean isAssignableToUnionType(BType source, BType target, Set<TypePair
                     }
                 }
                 // readonly can match to a union similar to any|error
-                if (sMember.tag == TypeTags.READONLY) {
+                if (sMember.tag == TypeTags.READONLY && isAssignable(symTable.anyAndReadonlyOrError, targetUnion)) {
                     unresolvedTypes.add(new TypePair(sMember, targetUnion));
```

Comment:
Do we still need to add this to unresolvedTypes?

Reply:
@KRVPerera mentioned that if we just remove that line, it may cause problems with cyclic types. I'm not sure if that's still the case, I'll check.